app/static/js/graphs/bubble-graph.js

```diff
@@ -130,11 +132,104 @@ function drawBubbleGraph(selector, data, replication) {
+               });
+
+                $(selectedBefore_pattern).remove();
+
+               //var total=$(`${selector} #bubbleData circle`)
+
+
+
+
+            var selectedBubbles = $(`${svg_selector} #bubbleData circle`).not(function(index, element) {
+                if(selected.includes(element.getAttribute("trait"))&& (controller == "delete") && (element.getAttribute("class").search(filter) == -1)){
+
+                    return false;}
+                 else if (selected.includes(element.getAttribute("trait")) && (controller == "add") && (element.getAttribute("class").search("disabled") == -1)){
+                   return false;}
+                else{ return true;}
+
+            });
+
+
+            for(i = 0; i < selectedBubbles.length; i++) {
+                var classAttr = $(selectedBubbles[i]).attr('class');
+                /*
+                if (classAttr.search("opaque") == -1){
+                    classAttr= classAttr +" opaque";
+                    window.alert("opaque added")
+                    }
+                    */
+                classAttr = [...new Set(classAttr.split(" "))].join(" ");
+
+                var cxAttr = $(selectedBubbles[i]).attr('cx');
+                //var cyAttr = $(selectedBubbles[i]).attr('cy');
+                //var rAttr = $(selectedBubbles[i]).attr('r');
+                var pubmedidAttr = $(selectedBubbles[i]).attr('pubmedid');
+                var authorAttr = $(selectedBubbles[i]).attr('author');
+                var accessionAttr = $(selectedBubbles[i]).attr('accession');
+                var nAttr = $(selectedBubbles[i]).attr('N');
+                var diseaseOrTraitAttr = $(selectedBubbles[i]).attr('DiseaseOrTrait');
+                var traitAttr = $(selectedBubbles[i]).attr('trait');
+                var cyAttr = yScale(nAttr);
+                var rAttr = sizeScale(nAttr);
+                var patternAttr=$(selectedBubbles[i]).attr('PatternSelection');
+
+                bubbleDataGroup.append("circle")
+                    .attr("class", classAttr)
+                    .attr("onclick", "circleClick(evt);")
+                    .attr("onmouseover", "circleMouseOver(evt);")
+                    .attr("cx", cxAttr)
+                    //.attr("cy", function (nAttr) { return yScale(nAttr); })
+                    //.attr("r", function(nAttr){ return sizeScale(nAttr) })
+                    .attr("cy", cyAttr)
+                    .attr("r", rAttr)
+                    .attr("pubmedid", pubmedidAttr)
+                    .attr("author", authorAttr)
+                    .attr("accession", accessionAttr)
+                    .attr("N", nAttr)
+                    .attr("DiseaseOrTrait", diseaseOrTraitAttr)
+                    .attr("trait", traitAttr)
+                    .attr("PatternSelection",patternAttr)
+                    .attr("ancestrySelection","ANCESTRY");
+                //$(selectedBubbles[i]).remove();
+
+            }
+
+        }
+
+

```

Comment:
Most of the code added here after `reDrawBubbleGraph` seems to be additional redraw code? Should it go inside `reDrawBubbleGraph`? It seems like its redrawing selected data, so perhaps it can go in its own function `redrawSelections`. Most of this redraw code is repeated below on lines 249-325. At this top level of abstraction its preferable to call functions that do the redrawing code, and any time code is repeated it should be refactored into a function.